File: python/tests/run_test.py
#!/usr/bin/env python3
"""
Test runner for Python implementation.
Accepts a test case JSON file and executes it.
"""

import json
import sys
import os
import re
import asyncio
from pathlib import Path
from typing import Dict, Any
import time

# Add parent directory to path to import timestep
sys.path.insert(0, str(Path(__file__).parent.parent))

from timestep import MultiModelProvider, MultiModelProviderMap, OllamaModelProvider

try:
    from agents import Agent, Runner, RunConfig, ModelSettings
    AGENTS_AVAILABLE = True
except ImportError:
    AGENTS_AVAILABLE = False


async def run_test(test_case: Dict[str, Any]) -> Dict[str, Any]:
    """Execute a test case and return the result."""
    start_time = time.time()
    
    try:
        # Setup provider based on test case
        setup = test_case.get("setup", {})
        provider_type = setup.get("provider_type", "multi")
        provider_config = setup.get("provider_config", {})
        
        if provider_type == "ollama":
            provider = OllamaModelProvider(
                api_key=provider_config.get("api_key"),
                base_url=provider_config.get("base_url"),
            )
        elif provider_type == "multi":
            provider_map = MultiModelProviderMap()
            if provider_config.get("ollama_api_key"):
                provider_map.add_provider(
                    "ollama",
                    OllamaModelProvider(api_key=provider_config.get("ollama_api_key"))
                )
            # Use environment variable if test config has placeholder, otherwise use config value
            openai_api_key = provider_config.get("openai_api_key")
            if openai_api_key == "test-key" or not openai_api_key:
                openai_api_key = os.environ.get("OPENAI_API_KEY", "")
            
            # Build provider kwargs with optional parameters
            provider_kwargs = {
                "provider_map": provider_map,
                "openai_api_key": openai_api_key,
            }
            
            # Add optional OpenAI provider options
            if provider_config.get("openai_base_url") is not None:
                provider_kwargs["openai_base_url"] = provider_config.get("openai_base_url")
            if provider_config.get("openai_organization") is not None:
                provider_kwargs["openai_organization"] = provider_config.get("openai_organization")
            if provider_config.get("openai_project") is not None:
                provider_kwargs["openai_project"] = provider_config.get("openai_project")
            if provider_config.get("openai_use_responses") is not None:
                provider_kwargs["openai_use_responses"] = provider_config.get("openai_use_responses")
            
            provider = MultiModelProvider(**provider_kwargs)
        else:
            raise ValueError(f"Unknown provider type: {provider_type}")
        
        # Get input
        test_input = test_case.get("input", {})
        model_name = test_input.get("model_name")
        run_agent = test_input.get("run_agent", False)
        stream = test_input.get("stream", False)
        
        actual_result = {
            "model_name": model_name,
            "provider_type": provider_type,
        }
        assertion_failures = []
        agent_output = None
        
        if run_agent:
            if not AGENTS_AVAILABLE:
                raise ImportError("agents package is required for agent tests. Install with: pip install openai-agents")
            
            # Create agent configuration
            agent_config = test_input.get("agent_config") or {}
            # Agent constructor requires 'name' parameter
            agent_kwargs = {
                "name": test_case.get("name", "test_agent"),
                "model": model_name
            }
            
            if agent_config and agent_config.get("system_prompt"):
                agent_kwargs["system"] = agent_config["system_prompt"]
            
            # Set temperature to 0 via ModelSettings
            temperature = agent_config.get("temperature", 0) if agent_config and agent_config.get("temperature") is not None else 0
            model_settings = ModelSettings(temperature=temperature)
            agent_kwargs["model_settings"] = model_settings
            
            # Create agent
            agent = Agent(**agent_kwargs)
            
            # Get user input
            user_input = test_input.get("user_input")
            if user_input is None:
                # Fallback to messages if user_input not provided
                messages = test_input.get("messages", [])
                if messages and len(messages) > 0:
                    first_msg = messages[0]
                    if isinstance(first_msg, dict):
                        user_input = first_msg.get("content", "Hello")
                    else:
                        user_input = str(first_msg) if first_msg else "Hello"
                else:
                    user_input = "Hello"
            
            # Convert user_input to string if it's a dict
            if isinstance(user_input, dict):
                user_input = user_input.get("content", str(user_input))
            elif not isinstance(user_input, str):
                user_input = str(user_input)
            
            # Create run config
            run_config = RunConfig(model_provider=provider)
            
            # Run agent (streaming or non-streaming)
            try:
                if stream:
                    # Use streaming mode - run_streamed returns a RunResultStreaming object
                    agent_output = ""
                    try:
                        stream_result = Runner.run_streamed(agent, user_input, run_config=run_config)
                        
                        # Try to get the stream from the result
                        # Based on error message, RunResultStreaming has 'stream_events' method
                        stream_iter = None
                        if hasattr(stream_result, 'stream_events'):
                            # stream_events is a method, so call it
                            stream_iter = stream_result.stream_events()
                        elif hasattr(stream_result, '__aiter__'):
                            stream_iter = stream_result
                        elif hasattr(stream_result, 'stream'):
                            stream_prop = stream_result.stream
                            # Check if it's a method or property
                            if callable(stream_prop):
                                stream_iter = stream_prop()
                            else:
                                stream_iter = stream_prop
                        elif hasattr(stream_result, 'events'):
                            events_prop = stream_result.events
                            if callable(events_prop):
                                stream_iter = events_prop()
                            else:
                                stream_iter = events_prop
                        elif callable(stream_result):
                            stream_iter = stream_result()
                        else:
                            # Try to get final_output as fallback
                            if hasattr(stream_result, 'final_output') and stream_result.final_output:
                                agent_output = str(stream_result.final_output)
                            else:
                                # Debug: print available attributes
                                attrs = [a for a in dir(stream_result) if not a.startswith('_')]
                                raise RuntimeError(f"Unable to iterate over RunResultStreaming: {type(stream_result)}, available attributes: {attrs}")
                        
                        # Iterate over the stream and collect text deltas
                        if stream_iter:
                            async for chunk in stream_iter:
                                # Extract delta/text from various possible structures
                                delta = None
                                
                                # Check for delta attribute
                                if hasattr(chunk, 'delta') and chunk.delta:
                                    delta = str(chunk.delta)
                                # Check for delta in dict
                                elif isinstance(chunk, dict):
                                    if 'delta' in chunk:
                                        delta = str(chunk['delta'])
                                    elif 'text' in chunk:
                                        delta = str(chunk['text'])
                                    elif 'content' in chunk:
                                        content = chunk['content']
                                        if isinstance(content, list):
                                            delta = ' '.join(str(c.get('text', c) if isinstance(c, dict) else c) for c in content)
                                        else:
                                            delta = str(content)
                                # Check for text attribute
                                elif hasattr(chunk, 'text') and chunk.text:
                                    delta = str(chunk.text)
                                # Check for content attribute
                                elif hasattr(chunk, 'content'):
                                    content = chunk.content
                                    if isinstance(content, list):
                                        delta = ' '.join(str(c.get('text', c) if isinstance(c, dict) else c) for c in content)
                                    else:
                                        delta = str(content)
                                
                                if delta:
                                    agent_output += delta
                            
                            # If we got nothing from deltas, try final_output
                            if not agent_output and hasattr(stream_result, 'final_output'):
                                final_output = stream_result.final_output
                                if final_output is not None:
                                    agent_output = str(final_output)
                                
                    except Exception as stream_error:
                        raise RuntimeError(f"Streaming failed: {str(stream_error)}") from stream_error
                    
                    # Ensure agent_output is not None or empty string "None"
                    if agent_output is None or agent_output == "None":
                        agent_output = ""
                else:
                    # Non-streaming mode
                    result = await Runner.run(agent, user_input, run_config=run_config)
                    
                    # Extract output from result (only for non-streaming)
                    # The result structure depends on the agents library version
                    if hasattr(result, 'final_output') and result.final_output:
                        # Try final_output first (most direct)
                        agent_output = str(result.final_output)
                    elif hasattr(result, 'messages') and result.messages:
                        # Get the last assistant message
                        last_message = result.messages[-1]
                        if hasattr(last_message, 'content'):
                            content = last_message.content
                            # content might be a list or string
                            if isinstance(content, list):
                                agent_output = ' '.join(str(c.get('text', c) if isinstance(c, dict) else c) for c in content)
                            else:
                                agent_output = str(content)
                        elif isinstance(last_message, dict):
                            agent_output = last_message.get("content", str(last_message))
                        else:
                            agent_output = str(last_message)
                    elif hasattr(result, 'content'):
                        agent_output = result.content
                    elif isinstance(result, str):
                        agent_output = result
                    else:
                        # Try to extract from string representation
                        result_str = str(result)
                        # Look for "Final output" pattern in Python RunResult
                        if "Final output" in result_str:
                            # Extract text between "Final output" and next section
                            match = re.search(r'Final output.*?:\s*(.+?)(?:\n-|\Z)', result_str, re.DOTALL)
                            if match:
                                agent_output = match.group(1).strip()
                            else:
                                agent_output = result_str
                        else:
                            agent_output = result_str
                
                if agent_output is not None and agent_output != "None":
                    actual_result["agent_output"] = agent_output
                
            except Exception as e:
                raise RuntimeError(f"Agent execution failed: {str(e)}") from e
        else:
            # Just verify model creation (original behavior)
            model = provider.get_model(model_name)
            # Model created successfully, no further action needed
        
        duration_ms = int((time.time() - start_time) * 1000)
        
        # Check expected results
        expected = test_case.get("expected") or {}
        
        # Validate model name
        if expected and expected.get("model_name") and model_name != expected.get("model_name"):
            assertion_failures.append({
                "assertion": {"field": "model_name", "operator": "equals", "value": expected.get("model_name")},
                "actual_value": model_name,
                "reason": f"Expected model_name {expected.get('model_name')}, got {model_name}"
            })
        
        # Validate agent output if provided
        if run_agent and agent_output is not None:
            agent_output_expectation = (expected.get("agent_output") or {}) if expected else {}
            
            # Check contains_text (case-insensitive)
            if agent_output_expectation.get("contains_text"):
                for text in agent_output_expectation["contains_text"]:
                    if text.lower() not in agent_output.lower():
                        assertion_failures.append({
                            "assertion": {"field": "agent_output", "operator": "contains", "value": text},
                            "actual_value": agent_output,
                            "reason": f"Expected agent output to contain '{text}', but it didn't"
                        })
            
            # Check excludes_text
            if agent_output_expectation.get("excludes_text"):
                for text in agent_output_expectation["excludes_text"]:
                    if text in agent_output:
                        assertion_failures.append({
                            "assertion": {"field": "agent_output", "operator": "excludes", "value": text},
                            "actual_value": agent_output,
                            "reason": f"Expected agent output to not contain '{text}', but it did"
                        })
            
            # Check min_length
            if agent_output_expectation.get("min_length") is not None:
                min_len = agent_output_expectation["min_length"]
                if len(agent_output) < min_len:
                    assertion_failures.append({
                        "assertion": {"field": "agent_output", "operator": "min_length", "value": min_len},
                        "actual_value": len(agent_output),
                        "reason": f"Expected agent output length >= {min_len}, got {len(agent_output)}"
                    })
            
            # Check max_length
            max_len = agent_output_expectation.get("max_length")
            if max_len is not None and max_len is not False:  # Handle None and False
                if len(agent_output) > max_len:
                    assertion_failures.append({
                        "assertion": {"field": "agent_output", "operator": "max_length", "value": max_len},
                        "actual_value": len(agent_output),
                        "reason": f"Expected agent output length <= {max_len}, got {len(agent_output)}"
                    })
            
            # Check exact_match (rarely used)
            if agent_output_expectation.get("exact_match"):
                expected_text = agent_output_expectation["exact_match"]
                if agent_output != expected_text:
                    assertion_failures.append({
                        "assertion": {"field": "agent_output", "operator": "equals", "value": expected_text},
                        "actual_value": agent_output,
                        "reason": f"Expected exact match, but got different output"
                    })
        
        status = "Passed" if not assertion_failures else "Failed"
        
        return {
            "test_name": test_case.get("name", "unknown"),
            "implementation": "python",
            "status": status,
            "duration_ms": duration_ms,
            "error": None,
            "actual_result": actual_result,
            "assertion_failures": assertion_failures,
        }
        
    except Exception as e:
        duration_ms = int((time.time() - start_time) * 1000)
        return {
            "test_name": test_case.get("name", "unknown"),
            "implementation": "python",
            "status": "Error",
            "duration_ms": duration_ms,
            "error": str(e),
            "actual_result": None,
            "assertion_failures": [],
        }


if __name__ == "__main__":
    if len(sys.argv) < 2:
        # Read from stdin
        test_case_json = sys.stdin.read()
    else:
        # Read from file
        with open(sys.argv[1], "r") as f:
            test_case_json = f.read()
    
    test_case = json.loads(test_case_json)
    result = asyncio.run(run_test(test_case))
    print(json.dumps(result))

File: typescript/tests/run_test.ts
#!/usr/bin/env node
/**
 * Test runner for TypeScript implementation.
 * Accepts a test case JSON file and executes it.
 */

import { readFileSync } from 'fs';
import { MultiModelProvider, MultiModelProviderMap, OllamaModelProvider } from '../timestep/index.ts';

let Agent: any;
let Runner: any;
let agentsAvailable = false;

// Load agents library dynamically
async function loadAgents() {
  if (agentsAvailable) return;
  
  try {
    // Try importing from @openai/agents
    // @ts-ignore - Dynamic import, may not be available
    const agentsModule = await import('@openai/agents');
    if (agentsModule.Agent && agentsModule.Runner) {
      Agent = agentsModule.Agent;
      Runner = agentsModule.Runner;
      agentsAvailable = true;
    } else {
      agentsAvailable = false;
    }
  } catch (e: any) {
    // Agents library not available - log for debugging
    console.error('Failed to load @openai/agents:', e.message);
    agentsAvailable = false;
  }
}

interface TestCase {
  name: string;
  setup: {
    provider_type: string;
    provider_config?: {
      openai_api_key?: string;
      openai_base_url?: string;
      openai_organization?: string;
      openai_project?: string;
      openai_use_responses?: boolean;
      ollama_api_key?: string;
      ollama_base_url?: string;
      [key: string]: any; // Allow other config options
    };
  };
  input: {
    model_name: string;
    messages?: Array<{ role: string; content: string }>;
    run_agent?: boolean;
    stream?: boolean;
    agent_config?: {
      system_prompt?: string;
      temperature?: number;
      max_tokens?: number;
      tools?: Array<any>;
    };
    user_input?: string | any;
  };
  expected?: {
    model_name?: string;
    provider_type?: string;
    should_succeed?: boolean;
    agent_output?: {
      contains_text?: string[];
      excludes_text?: string[];
      min_length?: number;
      max_length?: number;
      exact_match?: string;
      matches_pattern?: string;
    };
  };
}

interface TestResult {
  test_name: string;
  implementation: string;
  status: 'Passed' | 'Failed' | 'Error' | 'Skipped';
  duration_ms: number;
  error: string | null;
  actual_result: any;
  assertion_failures: Array<{
    assertion: any;
    actual_value: any;
    reason: string;
  }>;
}

async function runTest(testCase: TestCase): Promise<TestResult> {
  const startTime = Date.now();

  try {
    // Load agents library if needed
    await loadAgents();
    // Setup provider based on test case
    const { setup, input, expected } = testCase;
    const providerType = setup.provider_type || 'multi';
    const providerConfig = setup.provider_config || {};

    let provider: MultiModelProvider | OllamaModelProvider;

    if (providerType === 'ollama') {
      provider = new OllamaModelProvider({
        apiKey: providerConfig.api_key,
        baseURL: providerConfig.base_url,
      });
    } else if (providerType === 'multi') {
      const providerMap = new MultiModelProviderMap();
      if (providerConfig.ollama_api_key) {
        providerMap.addProvider(
          'ollama',
          new OllamaModelProvider({ apiKey: providerConfig.ollama_api_key })
        );
      }
      // Use environment variable if test config has placeholder, otherwise use config value
      let openaiApiKey = providerConfig.openai_api_key || '';
      if (openaiApiKey === 'test-key' || !openaiApiKey) {
        // Try to get from environment (works in both Node.js and Deno)
        if (typeof process !== 'undefined' && process.env) {
          openaiApiKey = process.env.OPENAI_API_KEY || '';
        } else {
          // @ts-ignore - Deno may not be available
          if (typeof Deno !== 'undefined' && Deno.env) {
            // @ts-ignore
            openaiApiKey = Deno.env.get('OPENAI_API_KEY') || '';
          }
        }
      }
      
      // Build provider options with optional parameters
      const providerOptions: any = {
        provider_map: providerMap,
        openai_api_key: openaiApiKey,
      };
      
      // Add optional OpenAI provider options
      if (providerConfig.openai_base_url !== undefined) {
        providerOptions.openai_base_url = providerConfig.openai_base_url;
      }
      if (providerConfig.openai_organization !== undefined) {
        providerOptions.openai_organization = providerConfig.openai_organization;
      }
      if (providerConfig.openai_project !== undefined) {
        providerOptions.openai_project = providerConfig.openai_project;
      }
      if (providerConfig.openai_use_responses !== undefined) {
        providerOptions.openai_use_responses = providerConfig.openai_use_responses;
      }
      
      provider = new MultiModelProvider(providerOptions);
    } else {
      throw new Error(`Unknown provider type: ${providerType}`);
    }

    // Get input
    const modelName = input.model_name;
    const runAgent = input.run_agent || false;
    const stream = input.stream || false;

    const actualResult: any = {
      model_name: modelName,
      provider_type: providerType,
    };
    let agentOutput: string | undefined;

    if (runAgent) {
      if (!agentsAvailable) {
        throw new Error(
          '@openai/agents package is required for agent tests. Install with: npm install @openai/agents'
        );
      }

      // Create agent configuration
      const agentConfig = input.agent_config || {};
      const agentOptions: any = { 
        name: testCase.name || 'test_agent',
        model: modelName 
      };

      if (agentConfig.system_prompt) {
        agentOptions.system = agentConfig.system_prompt;
      }

      // Set temperature to 0 via ModelSettings (plain object, not a class)
      const temperature = agentConfig.temperature !== undefined ? agentConfig.temperature : 0;
      agentOptions.model_settings = { temperature };

      // Create agent
      const agent = new Agent(agentOptions);

      // Get user input
      let userInput = input.user_input;
      if (userInput === undefined) {
        // Fallback to messages if user_input not provided
        const messages = input.messages || [];
        userInput = messages[0]?.content || 'Hello';
      }

      // Convert user_input to string if it's an object
      if (typeof userInput === 'object') {
        userInput = userInput.content || JSON.stringify(userInput);
      } else if (typeof userInput !== 'string') {
        userInput = String(userInput);
      }

      // Create runner
      const runner = new Runner({ modelProvider: provider });

      // Run agent (streaming or non-streaming)
      try {
        const result = await runner.run(agent, userInput, { stream });

        // Extract output from result
        if (stream) {
          // For streaming, result is an async iterable of stream events
          agentOutput = '';
          for await (const chunk of result) {
            // Extract text deltas from streaming events
            // Check multiple possible event types and structures
            const eventType = chunk?.type || chunk?.data?.type;
            
            // Handle output_text_delta events
            if (eventType === 'output_text_delta' || chunk?.type === 'output_text_delta') {
              // Extract delta from various possible locations
              let delta: string | undefined;
              
              // Direct delta property
              if (chunk?.delta && typeof chunk.delta === 'string') {
                delta = chunk.delta;
              }
              // Nested in data
              else if (chunk?.data?.delta && typeof chunk.data.delta === 'string') {
                delta = chunk.data.delta;
              }
              // In data.event
              else if (chunk?.data?.event?.delta && typeof chunk.data.event.delta === 'string') {
                delta = chunk.data.event.delta;
              }
              // In providerData
              else if (chunk?.providerData?.delta && typeof chunk.providerData.delta === 'string') {
                delta = chunk.providerData.delta;
              }
              
              if (delta) {
                agentOutput += delta;
              }
            }
            // Handle raw_model_stream_event with output_text_delta
            else if (eventType === 'raw_model_stream_event' || chunk?.type === 'raw_model_stream_event') {
              const data = chunk.data || chunk;
              if (data?.type === 'output_text_delta' && data.delta) {
                agentOutput += data.delta;
              } else if (data?.event?.type === 'output_text_delta' && data.event.delta) {
                agentOutput += data.event.delta;
              }
            }
            // Also check for final output in response_done events (as fallback)
            else if (eventType === 'response_done' || chunk?.type === 'response_done') {
              const response = chunk.response || chunk.data?.response;
              if (response?.output && Array.isArray(response.output) && response.output.length > 0) {
                const lastItem = response.output[response.output.length - 1];
                if (lastItem?.content && Array.isArray(lastItem.content)) {
                  const textParts = lastItem.content
                    .filter((c: any) => c.type === 'output_text' && c.text)
                    .map((c: any) => c.text);
                  if (textParts.length > 0) {
                    // Use final text if we don't have deltas yet, or as fallback
                    if (!agentOutput) {
                      agentOutput = textParts.join('');
                    }
                  }
                }
              }
            }
          }
        } else {
          // Non-streaming: extract from result object
          // The result structure depends on the agents library version
          if (result?.messages && Array.isArray(result.messages) && result.messages.length > 0) {
            // Get the last assistant message
            const lastMessage = result.messages[result.messages.length - 1];
            if (lastMessage?.content) {
              // content might be an array or string
              if (Array.isArray(lastMessage.content)) {
                agentOutput = lastMessage.content
                  .map((c: any) => c.text || c.content || JSON.stringify(c))
                  .join('');
              } else {
                agentOutput = lastMessage.content;
              }
            } else if (typeof lastMessage === 'string') {
              agentOutput = lastMessage;
            } else {
              agentOutput = JSON.stringify(lastMessage);
            }
          } else if (result?.content) {
            agentOutput = result.content;
          } else if (result?.state?.currentStep?.output) {
            // Try to extract from state.currentStep.output
            agentOutput = result.state.currentStep.output;
          } else if (result?.state?.lastModelResponse?.output_text) {
            // Try to extract from state.lastModelResponse.output_text
            agentOutput = result.state.lastModelResponse.output_text;
          } else if (typeof result === 'string') {
            agentOutput = result;
          } else {
            // Last resort: stringify, but try to extract text if it's a JSON string
            const str = JSON.stringify(result);
            try {
              const parsed = JSON.parse(str);
              if (parsed.state?.currentStep?.output) {
                agentOutput = parsed.state.currentStep.output;
              } else if (parsed.state?.lastModelResponse?.output_text) {
                agentOutput = parsed.state.lastModelResponse.output_text;
              } else {
                agentOutput = str;
              }
            } catch {
              agentOutput = str;
            }
          }
        }

        actualResult.agent_output = agentOutput;
      } catch (e: any) {
        throw new Error(`Agent execution failed: ${e.message || String(e)}`);
      }
    } else {
      // Just verify model creation (original behavior)
      const model = await provider.getModel(modelName);
      // Model created successfully, no further action needed
    }

    const durationMs = Date.now() - startTime;

    // Check expected results
    const assertionFailures: TestResult['assertion_failures'] = [];

    // Validate model name
    if (expected?.model_name && modelName !== expected.model_name) {
      assertionFailures.push({
        assertion: { field: 'model_name', operator: 'equals', value: expected.model_name },
        actual_value: modelName,
        reason: `Expected model_name ${expected.model_name}, got ${modelName}`,
      });
    }

    // Validate agent output if provided
    if (runAgent && agentOutput !== undefined) {
      const agentOutputExpectation = expected?.agent_output;

      if (agentOutputExpectation) {
        // Check contains_text (case-insensitive)
        if (agentOutputExpectation.contains_text) {
          for (const text of agentOutputExpectation.contains_text) {
            if (!agentOutput.toLowerCase().includes(text.toLowerCase())) {
              assertionFailures.push({
                assertion: { field: 'agent_output', operator: 'contains', value: text },
                actual_value: agentOutput,
                reason: `Expected agent output to contain '${text}', but it didn't`,
              });
            }
          }
        }

        // Check excludes_text (case-insensitive)
        if (agentOutputExpectation.excludes_text) {
          for (const text of agentOutputExpectation.excludes_text) {
            if (agentOutput.toLowerCase().includes(text.toLowerCase())) {
              assertionFailures.push({
                assertion: { field: 'agent_output', operator: 'excludes', value: text },
                actual_value: agentOutput,
                reason: `Expected agent output to not contain '${text}', but it did`,
              });
            }
          }
        }

        // Check min_length
        if (agentOutputExpectation.min_length !== undefined) {
          const minLen = agentOutputExpectation.min_length;
          if (agentOutput.length < minLen) {
            assertionFailures.push({
              assertion: { field: 'agent_output', operator: 'min_length', value: minLen },
              actual_value: agentOutput.length,
              reason: `Expected agent output length >= ${minLen}, got ${agentOutput.length}`,
            });
          }
        }

        // Check max_length
        if (agentOutputExpectation.max_length !== undefined && agentOutputExpectation.max_length !== null) {
          const maxLen = agentOutputExpectation.max_length;
          if (agentOutput.length > maxLen) {
            assertionFailures.push({
              assertion: { field: 'agent_output', operator: 'max_length', value: maxLen },
              actual_value: agentOutput.length,
              reason: `Expected agent output length <= ${maxLen}, got ${agentOutput.length}`,
            });
          }
        }

        // Check exact_match (rarely used)
        if (agentOutputExpectation.exact_match) {
          const expectedText = agentOutputExpectation.exact_match;
          if (agentOutput !== expectedText) {
            assertionFailures.push({
              assertion: { field: 'agent_output', operator: 'equals', value: expectedText },
              actual_value: agentOutput,
              reason: 'Expected exact match, but got different output',
            });
          }
        }
      }
    }

    const status = assertionFailures.length === 0 ? 'Passed' : 'Failed';

    return {
      test_name: testCase.name,
      implementation: 'typescript',
      status,
      duration_ms: durationMs,
      error: null,
      actual_result: actualResult,
      assertion_failures: assertionFailures,
    };
  } catch (error: any) {
    const durationMs = Date.now() - startTime;
    return {
      test_name: testCase.name,
      implementation: 'typescript',
      status: 'Error',
      duration_ms: durationMs,
      error: error.message || String(error),
      actual_result: null,
      assertion_failures: [],
    };
  }
}

// Main execution
const testCaseFile = process.argv[2];
if (!testCaseFile) {
  console.error('Usage: run_test.ts <test_case.json>');
  process.exit(1);
}

const testCaseJson = readFileSync(testCaseFile, 'utf-8');
const testCase: TestCase = JSON.parse(testCaseJson);

runTest(testCase)
  .then((result) => {
    console.log(JSON.stringify(result));
  })
  .catch((error) => {
    console.error(JSON.stringify({
      test_name: testCase.name || 'unknown',
      implementation: 'typescript',
      status: 'Error',
      duration_ms: 0,
      error: error.message || String(error),
      actual_result: null,
      assertion_failures: [],
    }));
    process.exit(1);
  });

